### ISSUE TITLE
 kotlin.native.concurrent.InvalidMutabilityException: mutation attempt of frozen kotlinx.coroutines.flow.StateFlowImpl

### ISSUE THREAD

**Issue by matt400m** [Kotlin/kotlinx.coroutines#2584]
**Kotlin**: 1.4.31
**Version**: 1.4.3-mt

After Kotlin update from 1.4.21 -> 1.4.31 I have a strange behaviour
in src/iosTest I have:

```kotlin
@Test
    fun testSingleThreadWithFreeze() {
        val f = MutableStateFlow(1)
        f.value = 2
        f.freeze()
        f.value = 3
        runBlocking {
            f.take(1).collect {
                println("got $it")
            }

            println("got ${f.value}")
        }
    }
```

and it's failed.

stacktrace:

```kotlin.native.concurrent.InvalidMutabilityException: mutation attempt of frozen kotlinx.coroutines.flow.StateFlowImpl@6e659b8
kotlin.native.concurrent.InvalidMutabilityException: mutation attempt of frozen kotlinx.coroutines.flow.StateFlowImpl@6e659b8
	at kotlin.Throwable#<init>(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/Throwable.kt:23)
	at kotlin.Exception#<init>(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/Exceptions.kt:23)
	at kotlin.RuntimeException#<init>(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/Exceptions.kt:34)
	at kotlin.native.concurrent.InvalidMutabilityException#<init>(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/concurrent/Freezing.kt:22)
	at <global>.ThrowInvalidMutabilityException(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/concurrent/Internal.kt:93)
	at <global>.MutationCheck(Unknown Source)
	at kotlinx.coroutines.flow.StateFlowImpl.<set-sequence>#internal(/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/flow/StateFlow.kt:263)
	at kotlinx.coroutines.flow.StateFlowImpl.updateState#internal(/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/flow/StateFlow.kt:284)
	at kotlinx.coroutines.flow.StateFlowImpl.<set-value>#internal(/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/flow/StateFlow.kt:268)
	at <global>.MutableFlowStateTest#testSingleThreadWithFreeze(/Users//projekty/-common-shared/Sources/shared/src/iosTest/kotlin/MutableFlowStateTest.kt:34)
	at $MutableFlowStateTest$test$0.$testSingleThreadWithFreeze$FUNCTION_REFERENCE$182.invoke#internal(/Users//projekty/-common-shared/Sources/shared/src/iosTest/kotlin/MutableFlowStateTest.kt:30)
	at $MutableFlowStateTest$test$0.$testSingleThreadWithFreeze$FUNCTION_REFERENCE$182.$<bridge-UNNN>invoke(/Users///-common-shared/Sources/shared/src/iosTest/kotlin/MutableFlowStateTest.kt:30)
	at kotlin.native.internal.test.BaseClassSuite.TestCase#run(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/TestSuite.kt:85)
	at kotlin.native.internal.test.TestRunner.run#internal(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:241)
	at kotlin.native.internal.test.TestRunner.runIteration#internal(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:265)
	at kotlin.native.internal.test.TestRunner#run(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/TestRunner.kt:280)
	at kotlin.native.internal.test#testLauncherEntryPoint(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:22)
	at kotlin.native.internal.test#main(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:26)
	at <global>.Konan_start(/Users/teamcity1/teamcity_work/11ac87a349af04d5/runtime/src/main/kotlin/kotlin/native/internal/test/Launcher.kt:25)
	at <global>.Init_and_run_start(Unknown Source)
	at <global>.start(Unknown Source)
	at <global>.0x0(Unknown Source)
```

But when I write ```f.value.freeze()``` I don't have an exception ```kotlin.native.concurrent.InvalidMutabilityException: ```

**working**:
```kotlin
 @Test
    fun testSingleThreadWithFreeze() {
        val f = MutableStateFlow(1)
        f.value = 2
        f.value.freeze()
        f.value = 3
        runBlocking {
            f.take(1).collect {
                println("got $it")
            }

            println("got ${f.value}")
        }
    }
```

**Comment by mvicsokolova** [comment]
Actually, the behaviour is correct for both cases, you've provided.

In the first case, freezing the `MutableStateFlow` makes the whole object graph deeply immutable, so mutation of the field `value` causes `InvalidMutabilityException`. While in the second case you do not freeze the whole object, but just the field.

**Comment by qwwdfsad** [comment]
It seems like it was some kind of a stuck update, in that scenario, it's typically worth to make full clean build and reset IDEA caches.

This very problem was fixed in `1.4.3-native-mt` and non-reproducible with Kotlin 1.5.0 and coroutines `1.5.0-native-mt` as well.
